src/backend/cdb/cdbdtxrecovery.c

```diff
@@ -153,6 +154,18 @@ recoverTM(void)
 	elogif(Debug_print_full_dtm, LOG,
 		   "Starting to Recover DTM...");
 
+	/*
+	 * We'd better terminate residual QE processes to avoid potential issues,
+	 * e.g. shared snapshot collision, etc. We do soft-terminate here so it is
+	 * still possible there are residual QE processes but it's better than doing
+	 * nothing.
+	 *
+	 * We just do this when there was abnormal shutdown on master or standby
+	 * promote, else mostly there should not have residual QE processes.
+	 */
+	if (Gp_role == GP_ROLE_DISPATCH)
+		TerminateMppBackends();
```

Comment:
Why did you decide to run the termination process unconditionally, despite xlog recovery? Why do we need to check role here?

Reply:
I revised the patch and fixed my mistake.